### PR TITLE
Ks 323 keystone evm integration tests

### PR DESCRIPTION
**Pull request by ettec**
https://smartcontract-it.atlassian.net/browse/KS-323

Framework for running end to end tests for keystone.   The test framework utilizes all keystone contracts and simulates running a trigger DON, workflow DON and target DON.   It achieves this by mocking out the dispatcher and libocr to simulate a multi DON environment within the testing process.  It uses the existing integration testing framework to instantiate nodes and uses the geth simulated backed to run all required contract code.

A couple of tests are implemented as part of this PR to test the framework and serve as examples. TBD Each tests takes in the region of 5s to run, possible this could also be improved by decreasing the block time and ocr protocol round time.

### CODE REVIEW

File: core/services/relay/evm/method_binding.go [hunk 1]
```diff
@@ -124,6 +124,8 @@ func (m *methodBinding) blockNumberFromConfidence(ctx context.Context, confidenc
 		return big.NewInt(finalized.Number), nil
 	} else if confirmations == evmtypes.Unconfirmed {
 		return big.NewInt(latest.BlockNumber()), nil
+	} else if confirmations == evmtypes.Latest {
```
Comment (ilija42) [inline core/services/relay/evm/method_binding.go]: I think that you can change `Unconfirmed` to use nil since `Latest` and `Unconfirmed` are the same thing here. The only difference is that right now the `latest` tag is used in `eth_getBlockByNumber` rpc call by head tracker, instead of in `CallContract`.

Comment (ettec) [reply core/services/relay/evm/method_binding.go]: Thanks for the suggestion, that's the change I've made

File: core/capabilities/integration_tests/mock_trigger.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/datastreams"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+const triggerID = "streams-trigger@1.0.0"
```
Comment (archseer) [inline core/capabilities/integration_tests/mock_trigger.go:16]: I actually have a similar thing here, I'm hoping we could just make this a new trigger type useful for testing/development in the core node: https://github.com/smartcontractkit/chainlink/pull/13716/files#diff-905de3fd1acdb9e2daa319880d9a528cca7f034d3b8920eec76905b7e6b52acd

File: core/capabilities/integration_tests/keystone_contracts_setup.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/feeds_consumer"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/forwarder"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/cltest"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+
+	//	helpers "github.com/smartcontractkit/chainlink/core/scripts/common"
```
Comment (krehermann) [inline core/capabilities/integration_tests/keystone_contracts_setup.go]: rm?

File: core/capabilities/integration_tests/keystone_contracts_setup.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+	remotetypes "github.com/smartcontractkit/chainlink/v2/core/capabilities/remote/types"
+	kcr "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/capabilities_registry"
+)
+
+type peer struct {
+	PeerID string
+	Signer string
+}
+
+func peerIDToB(peerID string) ([32]byte, error) {
```
Comment (krehermann) [inline core/capabilities/integration_tests/keystone_contracts_setup.go]: nit/ `Bytes` is clearer than `B` as a suffix

File: core/capabilities/integration_tests/keystone_contracts_setup.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+	transactOpts *bind.TransactOpts, backend *ethBackend) common.Address {
+	addr, _, reg, err := kcr.DeployCapabilitiesRegistry(transactOpts, backend)
+	require.NoError(t, err)
+
+	backend.Commit()
+
+	streamsTrigger := kcr.CapabilitiesRegistryCapability{
+		LabelledName:   "streams-trigger",
+		Version:        "1.0.0",
+		CapabilityType: uint8(0), // trigger
```
Comment (krehermann) [inline core/capabilities/integration_tests/keystone_contracts_setup.go]: can we import and use public consts or enums?

File: core/capabilities/integration_tests/keystone_contracts_setup.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	// workflow DON
+	ps, err := peers(workflowDonPeers)
+	require.NoError(t, err)
+
+	cfgs := []kcr.CapabilitiesRegistryCapabilityConfiguration{
+		{
+			CapabilityId: ocrid,
+		},
+	}
+	_, err = reg.AddDON(transactOpts, ps, cfgs, false, true, 2)
```
Comment (krehermann) [inline core/capabilities/integration_tests/keystone_contracts_setup.go]: do it have to be `2`? if so please add a comment or assign var with semantic meaning

File: core/capabilities/integration_tests/streams_test.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+func Test_AllAtOnceTransmissionSchedule(t *testing.T) {
+	ctx := testutils.Context(t)
+
+	// The don IDs set in the below calls are inferred from the order in which the dons are added to the capabilities registry
+	// in the setupCapabilitiesRegistryContract function, should this order change the don IDs will need updating.
+	workflowDonInfo := createDonInfo(t, 1, 5, 1)
+	triggerDonInfo := createDonInfo(t, 2, 7, 1)
+	targetDonInfo := createDonInfo(t, 3, 4, 1)
```
Comment (cedric-cordenier) [inline core/capabilities/integration_tests/streams_test.go]: @ettec Maybe passing in a struct would be better here? That way we'd know what each of the ints is for: I assume it's donID, numNodes and F?

File: core/capabilities/integration_tests/streams_test.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		createFeedReport(t, big.NewInt(3), 7, feedIDs[1], triggerDonInfo.keyBundles),
+		createFeedReport(t, big.NewInt(2), 6, feedIDs[2], triggerDonInfo.keyBundles),
+	}
+
+	triggerSink.sendReports(reports)
+
+	waitForConsumerReports(t, consumer, reports)
+}
+
+func setupStreamDonsWithTransmissionSchedule(ctx context.Context, t *testing.T, workflowDonInfo donInfo, triggerDonInfo donInfo, targetDonInfo donInfo,
```
Comment (cedric-cordenier) [inline core/capabilities/integration_tests/streams_test.go]: It looks like this totally bypassed the syncer? Is that right? I think it would be awesome if we could include that (can be done in a follow up though! as this PR is quite large already).

Comment (ettec) [reply core/capabilities/integration_tests/streams_test.go]: It doesn't bypass the syncer, that is being used as it would during normal run to setup the capabilities.  Perhaps you mean the transmission schedule?  Pre your recent changes, this is/was being passed in on the workflow, which is what's happening here.

Comment (ettec) [reply core/capabilities/integration_tests/streams_test.go]: The syncer is being used to instantiate all capabilities apart from the trigger capability as the test needs a handle on this to inject the test messages.  Though the syncer is still setting up the remote shims for trigger.

**Review by cedric-cordenier — DISMISSED** [review]

File: core/capabilities/integration_tests/streams_test.go (hunk 7 above)
Comment (ettec) [reply core/capabilities/integration_tests/streams_test.go]: sorted, ta

File: core/capabilities/integration_tests/mock_trigger.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+type reportsSink struct {
+	triggers []streamsTrigger
+}
+
+func (r *reportsSink) sendReports(reportList []*datastreams.FeedReport) {
+	for _, trigger := range r.triggers {
+		resp, err := wrapReports(reportList, "1", 12, datastreams.SignersMetadata{})
```
Comment (DeividasK) [inline core/capabilities/integration_tests/mock_trigger.go:48]: Should the signers metadata include signatures? Otherwise, how are we validating the reports further down the line?

File: core/capabilities/integration_tests/setup.go [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+			report := feedToReport[feedID]
+			decodedReport, err := reporttypes.Decode(report.FullReport)
+			require.NoError(t, err)
+			assert.Equal(t, decodedReport.BenchmarkPrice, feed.Price)
+			assert.Equal(t, decodedReport.ObservationsTimestamp, feed.Timestamp)
+
+			feedCount++
+			if feedCount == len(triggerFeedReports) {
+				return
+			}
```
Comment (DeividasK) [inline core/capabilities/integration_tests/setup.go]: Make the error explicit with "expected reports x,y,z; received ..."

File: core/capabilities/integration_tests/streams_test.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+	reports := []*datastreams.FeedReport{
+		createFeedReport(t, big.NewInt(1), 5, feedIDs[0], triggerDonInfo.keyBundles),
+		createFeedReport(t, big.NewInt(3), 7, feedIDs[1], triggerDonInfo.keyBundles),
+		createFeedReport(t, big.NewInt(2), 6, feedIDs[2], triggerDonInfo.keyBundles),
+	}
+
+	triggerSink.sendReports(reports)
+
+	waitForConsumerReports(t, consumer, reports)
```
Comment (DeividasK) [inline core/capabilities/integration_tests/streams_test.go]: This is an assertion function, so could probalby stay in the test file :man_shrugging:

File: core/capabilities/integration_tests/workflow.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+	for _, feedID := range feedIDs {
+		triggerFeedIDs += fmt.Sprintf("        - \"%s\"\n", feedID)
+	}
+
+	aggregationFeeds := ""
+	for _, feedID := range feedIDs {
+		aggregationFeeds += fmt.Sprintf("          \"%s\":\n            deviation: \"0.001\"\n            heartbeat: 3600\n", feedID)
+	}
+
+	workflowJobSpec := testspecs.GenerateWorkflowJobSpec(t, fmt.Sprintf(hardcodedWorkflow, workflowName, workflowOwner, triggerFeedIDs, aggregationFeeds,
```
Comment (DeividasK) [inline core/capabilities/integration_tests/workflow.go:69]: Keith had something for this in chainlink-common?

File: core/services/relay/evm/method_binding.go [hunk 13]
```diff
@@ -115,15 +115,15 @@ func (m *methodBinding) blockNumberFromConfidence(ctx context.Context, confidenc
...
 	if err != nil {
 		return nil, err
 	}
 
 	if confirmations == evmtypes.Finalized {
 		return big.NewInt(finalized.Number), nil
 	} else if confirmations == evmtypes.Unconfirmed {
-		return big.NewInt(latest.BlockNumber()), nil
+		return nil, nil
 	}
```
Comment (DeividasK) [inline core/services/relay/evm/method_binding.go]: Could we do this in a separate PR just so that Ilja / relayer owner could stamp it?

File: core/capabilities/integration_tests/workflow.go (hunk 12 above)
Comment (krehermann) [reply core/capabilities/integration_tests/workflow.go:69]: there is this, but i don't think it actually suits the need here. and coupling to common isn't worth the hassle for this functionality
https://github.com/smartcontractkit/chainlink-common/blob/ea76a8d41d877f98f4b56193b7aeba95e35bf71f/pkg/workflows/testutils.go#L12

File: core/services/relay/evm/method_binding.go (hunk 13 above)
Comment (ettec) [reply core/services/relay/evm/method_binding.go]: done here-> https://github.com/smartcontractkit/chainlink/pull/13906

**Review by cedric-cordenier — DISMISSED** [review]

**Review by DeividasK — DISMISSED** [review]

**Review by cedric-cordenier — APPROVED** [review]

**Review by DeividasK — APPROVED** [review]
